validator/Dockerfile

```diff
@@ -0,0 +1,67 @@
+# Copyright 2018 Cargill Incorporated
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# docker build -f /Dockerfile -t sawtooth-validator-local .
+
+# -------------===  build ===-------------
+
+FROM ubuntu:xenial
+
+RUN echo "deb http://repo.sawtooth.me/ubuntu/ci xenial universe" >> /etc/apt/sources.list \
+ && apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 8AA7AF1F1091A5FD \
+ && apt-get update
+
+RUN apt-get install -y -q --allow-downgrades \
+    git \
+    python3 \
+    python3-stdeb
+
+RUN apt-get install -y -q --allow-downgrades \
+    python3-grpcio \
+    python3-grpcio-tools \
+    python3-protobuf
+
+RUN apt-get install -y -q --allow-downgrades \
+    python3-cbor \
+    python3-colorlog \
+    python3-dev \
+    python3-lmdb \
+    python3-netifaces=0.10.4-0.1build2 \
+    python3-pyformance \
+    python3-secp256k1 \
+    python3-toml \
+    python3-yaml \
+    python3-zmq \
+    unzip
+
+RUN curl -OLsS https://github.com/google/protobuf/releases/download/v3.5.1/protoc-3.5.1-linux-x86_64.zip \
+ && unzip protoc-3.5.1-linux-x86_64.zip -d protoc3 \
+ && rm protoc-3.5.1-linux-x86_64.zip
+
+RUN curl https://sh.rustup.rs -sSf > /usr/bin/rustup-init \
+ && chmod +x /usr/bin/rustup-init \
+ && rustup-init -y
+
+ENV PATH=$PATH:/project/sawtooth-core/bin:/protoc3/bin:/project/sawtooth-core/bin:/root/.cargo/bin \
```

Comment:
Why don't we build the cargo dependencies here. That way you don't have to on every `up`. It can cut like 90 seconds off of the process.

Supply chain does it here:
https://github.com/hyperledger/sawtooth-supply-chain/blob/master/processor/Dockerfile#L35

Reply:
The "mounted" images should not COPY any source code nor should they have to be rebuilt when the source changes. They should only have to be rebuilt when dependencies change. I think a better solution to this problem is to use a persistent volume that you mount when you run the container. That way you aren't caching possibly outdated manifest files or dependencies in the image itself. Something like the following if you are using plain docker:
```
docker volume create cargo-registry && docker volume create validator-target;
docker run -it -v cargo-registry:/root/.cargo/registry -v validator-target:/project/sawtooth-core/validator/target sawtooth-validator
```

Reply:
I don't see how including dependencies in the form of a `COPY`ed `Cargo.toml` is functionally different than including dependencies as hard-coded `apt-get` commands. Either way, if the dependency changes, the docker image will need to be rebuilt.

Also, it is worth pointing out that _not_ rebuilding the Dockerfile will _not_ break the build. The out-of-date cached dependencies will simply be installed when you `cargo build` during `up`. If you do not build when building the image, you will have to install _every_ dependency on `up`, so I feel like installing the dependencies in the image still has the advantage here.

I am not familiar with the commands you listed. If they make it so you can _just_ build your changes when you `up`, and not have to reinstall and rebuild all of your Rust dependencies, then I'm down.